File: nstableheadercellbugobjc/ViewController.m
//
//  ViewController.m
//  nstableheadercellbugobjc
//
//  Created by omochimetaru on 2017/11/21.
//  Copyright © 2017年 omochimetaru. All rights reserved.
//

#import "ViewController.h"

@interface MyHeaderCell : NSTableHeaderCell {}

@end

@implementation MyHeaderCell

- (void)dealloc {
    NSLog(@"dealloc %p", self);
}

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    for (NSTableColumn * column in _tableView.tableColumns) {
        MyHeaderCell * cell = [[MyHeaderCell alloc] initTextCell:column.title];
        column.headerCell = cell;
    }

}

@end

File: nstableheadercellbug/ViewController.swift
//
//  ViewController.swift
//  nstableheadercellbug
//
//  Created by omochimetaru on 2017/11/21.
//

import Cocoa

func addr(_ obj: AnyObject) -> UnsafeMutableRawPointer {
    return Unmanaged.passUnretained(obj).toOpaque()
}

class Cat {
    var name: String = "nyaa"
    
    init() {
        print("[Cat(\(addr(self)))::init]")
    }
    
    deinit {
        print("[Cat(\(addr(self)))::deinit]")
    }
}

class MyHeaderCell : NSTableHeaderCell {
    var cat: Cat = .init()

    required init(coder: NSCoder) {
        super.init(coder: coder)
        print("[MyHeaderCell(\(addr(self)))::init(coder:)]")
    }
    
    override init(textCell string: String) {
        super.init(textCell: string)
        print("[MyHeaderCell(\(addr(self)))::init(textCell:)]")
    }
    
    deinit {
        print("[MyHeaderCell(\(addr(self)))::deinit]")
    }
}

class ViewController: NSViewController {
    @IBOutlet weak var tableView: NSTableView!

    override func viewDidLoad() {
        super.viewDidLoad()
        
        for column in tableView.tableColumns {
            let headerCell = MyHeaderCell.init(textCell: column.title)
            column.headerCell = headerCell
        }
    }
}

